### PR TITLE
Check IOException for corruption specific messages

### LINKED ISSUE #34322: [CI] RemoveCorruptedShardDataCommandTests testCorruptedIndex fails
logs: https://www.google.com/url?q=https%3A%2F%2Felasticsearch-ci.elastic.co%2Fjob%2Felastic%2Belasticsearch%2B6.x%2Bmultijob-unix-compatibility%2Fos%3Damazon%2F42%2Fconsole&sa=D&sntz=1&usg=AFQjCNFFG4hg9naqoVYMqXTq8OD-lszrlw

It is reproducible with:
```
./gradlew :server:test -Dtests.seed=F13528ED31F73FE7 -Dtests.class=org.elasticsearch.index.shard.RemoveCorruptedShardDataCommandTests -Dtests.method="testCorruptedIndex" -Dtests.security.manager=true -Dtests.locale=es-CO -Dtests.timezone=Canada/Central -Dcompiler.java=11 -Druntime.java=8
```

The reason:
corrupting segments file, when Lucene reads it, it throws generic IOException at https://github.com/apache/lucene-solr/blob/1d85cd783863f75cea133fb9c452302214165a4d/lucene/core/src/java/org/apache/lucene/store/DataInput.java#L141 and that's why ES would not consider it as an index corruption and does not catch it at https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/store/Store.java#L188 and therefore would not create a corruption marker (while RemoveCorruptedShardDataCommand relies on its presence) 

It would be better if Lucene throws more specific exception rather generic one.

**update**: https://issues.apache.org/jira/browse/LUCENE-8525 to address this problem

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
Lucene throws generic IOException due to corrupted data - https://github.com/apache/lucene-solr/blob/1d85cd783863f75cea133fb9c452302214165a4d/lucene/core/src/java/org/apache/lucene/store/DataInput.java#L141 - it is not handled properly on its level and effectively SegmentInfo.readCommit violates its contract 
```
   * @throws CorruptIndexException if the index is corrupt
   * @throws IOException if there is a low-level IO error
   */
  public static final SegmentInfos readCommit(Directory directory, String segmentFileName) throws IOException {
```

While there is https://issues.apache.org/jira/browse/LUCENE-8525 to address this problem, we can prevent this problem on our side relying on an exception message like we do in https://github.com/elastic/elasticsearch/blob/237650e9c054149fd08213b38a81a3666c1868e5/server/src/main/java/org/elasticsearch/common/transport/NetworkExceptionHelper.java#L40

**Note**: this PR is more a placeholder if we are not able to manage Lucene issue in some reasonable time

Closes #34322

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/store/Store.java [hunk 1]
```diff
@@ -203,6 +203,11 @@ private static SegmentInfos readSegmentsInfo(IndexCommit commit, Directory direc
             // TODO this should be caught by lucene - EOF is almost certainly an index corruption
             throw new CorruptIndexException("Read past EOF while reading segment infos", "commit(" + commit + ")", eof);
         } catch (IOException exception) {
+            final String message = exception.getMessage();
+            // DataInput throws a generic IOException on odd data cases
+            if (message != null && (message.startsWith("Invalid vLong detected") || message.startsWith("Invalid vInt detected"))) {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/store/Store.java]: If we are going to do this, we should add a test that asserts these exception messages continue to be generated by Lucene (i.e., that we are not prone to them silently changing underneath us).

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a comment about adding a test.

File: server/src/main/java/org/elasticsearch/index/store/Store.java (hunk 1 above)
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/index/store/Store.java]: @jasontedor I've addressed that in 2a18280

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Pull request closed without merging**